x/liquidstaking/keeper/liquidstaking.go

```diff
@@ -1352,8 +1350,11 @@ func (k Keeper) IsEnoughToCoverSlash(ctx sdk.Context, insurance types.Insurance)
 }
 
 // GetAvailableChunkSlots returns a number of chunk which can be paired.
-func (k Keeper) GetAvailableChunkSlots(ctx sdk.Context) sdk.Int {
-	return k.MaxPairedChunks(ctx).Sub(sdk.NewInt(k.getNumPairedChunks(ctx)))
+func (k Keeper) GetAvailableChunkSlots(ctx sdk.Context, u, uHardCap sdk.Dec) sdk.Int {
```

Comment:
When U is greater than hardcap, remainingU and result can be negative, so safe logic seems to be needed 

Also, for easy unit test addition, I think it would be good to get a supply as a arg and refactoring with a type level function rather than a keeper level

If the above refactoring is applied to other Calc* functions(e.g. CalcDynamicFeeRate, CalcUtilizationRatio, etc), it is likely to reduce duplicate state get and use type level testing

Reply:
Sounds good! I'll try refactoring it!

Reply:
done! (https://github.com/b-harvest/Canto/pull/18/commits/6d3baec5dffd63257c54b81bf1e7c59c7ab6468d)
did refactoring and added safe logic too.

Reply:
Since type level GetAvailableChunkSlots has been added, I think you can delete this duplicated function

Reply:
done 👍  https://github.com/b-harvest/Canto/commit/0dd25100f7b881a83863b8c99c4b47dd8a2d8ba8